src/web/api.rs

```diff
@@ -97,9 +211,328 @@ async fn get_players(
+            .dispatch()
+            .await;
+
+        assert_eq!(rocket::http::Status::Ok, resp.status(),);
+        let parsed = parse_result::<Vec<ApiBracket>>(&resp.into_string().await.unwrap())?
+            .map_err(|e| anyhow!("{e}"))?;
+        assert_eq!(2, parsed.len());
+        Ok(())
+    }
+    #[tokio::test]
+    async fn test_get_races() -> anyhow::Result<()> {
+        let c = setup().await?;
+        let s = run_with_db(&c, |db| {
+            let ns = NewSeason::new("Any% NMG", "alttp", "Any% NMG", db)?.save(db)?;
+            let b = NewBracket::new(&ns, "bracket 1", BracketType::Swiss).save(db)?;
+            let round = NewBracketRound::new(&b, 1).save(db)?;
+            let p1 = NewPlayer::new("p1", "1", None, None).save(db)?;
+            let p2 = NewPlayer::new("p2", "2", None, None).save(db)?;
+            NewPlayerBracketEntry::new(&b, &p1).save(db)?;
+            NewPlayerBracketEntry::new(&b, &p2).save(db)?;
+            bracket_races::insert_bulk(&vec![NewBracketRace::new(&b, &round, &p1, &p2)], db)?;
+            Ok(ns)
+        })
+        .await?;
+
+        let unfiltered_resp = c
+            .get(format!("/api/v1/season/{}/races", s.ordinal))
+            .dispatch()
+            .await;
+        assert_eq!(rocket::http::Status::Ok, unfiltered_resp.status(),);
+        let parsed = parse_result::<Vec<ApiRace>>(&unfiltered_resp.into_string().await.unwrap())?
+            .map_err(|e| anyhow!("{e}"))?;
+        assert_eq!(1, parsed.len());
+
+        let bad_state = c
+            .get(format!("/api/v1/season/{}/races?state=foo", s.ordinal))
+            .dispatch()
+            .await;
+        assert_eq!(rocket::http::Status::Ok, bad_state.status(),);
+        let parsed = parse_result::<Vec<ApiRace>>(&bad_state.into_string().await.unwrap())?;
+        assert!(parsed.is_err());
+        assert_eq!("Bad Request", parsed.err().unwrap());
+
+        // `urlencoding::encode()` urlencodes the `=` sign!
+        let new_resp = c
+            .get(format!(
+                "/api/v1/season/{}/races?state={}",
+                s.ordinal,
+                urlencoding::encode(r#""New""#)
+            ))
+            .dispatch()
+            .await;
+        assert_eq!(rocket::http::Status::Ok, new_resp.status(),);
+        let parsed = parse_result::<Vec<ApiRace>>(&new_resp.into_string().await.unwrap())?
+            .map_err(|e| anyhow!("{e}"))?;
+        assert_eq!(1, parsed.len());
+
+        let scheduled_resp = c
+            .get(format!(
+                "/api/v1/season/{}/races?state={}",
+                s.ordinal,
+                urlencoding::encode(r#""Scheduled""#)
+            ))
+            .dispatch()
+            .await;
+        assert_eq!(rocket::http::Status::Ok, scheduled_resp.status(),);
+        let parsed = parse_result::<Vec<ApiRace>>(&scheduled_resp.into_string().await.unwrap())?
+            .map_err(|e| anyhow!("{e}"))?;
+        assert_eq!(0, parsed.len());
+        Ok(())
+    }
+}
```

Comment:
there's room for more thoroughness but i think this is fine. it's for sure the best tested module now :\